src/eval_utils/run.ts

```diff
@@ -0,0 +1,725 @@
+
+            if ("flow" in request) {
+                if (!_.isEqual(state!.evaluatedVersion, request.flow)) {
+                    response = await originalLog(
+                        {
+                            ...request,
+                            // @ts-ignore Log under the version expected by evaluation, not
+                            // one determined by decorators. Otherwise the evaluation will stale
+                            flow: state?.evaluatedVersion,
+                            output: undefined,
+                            error: `The version of the evaluated Flow must match the version of the callable. Expected: ${JSON.stringify(state!.evaluatedVersion)}, got: ${JSON.stringify(request.flow)}`,
+                        },
+                        options,
+                    );
+                }
+            }
+
+            if ("prompt" in request) {
+                if (!_.isEqual(state!.evaluatedVersion, request.prompt)) {
+                    response = await originalLog({
+                        ...request,
+                        // @ts-ignore Log under the version expected by evaluation, not
+                        // one determined by decorators. Otherwise the evaluation will stale
+                        prompt: state?.evaluatedVersion,
+                        output: undefined,
+                        error: `The version of the evaluated Prompt must match the version of the callable. Expected: ${JSON.stringify(state!.evaluatedVersion)}, got: ${JSON.stringify(request.prompt)}`,
+                    });
+                }
+            }
+
+            if (response === undefined) {
+                // Version validation passed, make a normal request
+                response = await originalLog(request, options);
+            }
+
+            // @ts-ignore
+            uploadCallback(response.id);
+        } else {
+            // @ts-ignore
+            response = await originalLog(request, options);
+        }
+
+        return response;
+    };
+
+    client.log = _overloadedLog.bind(client);
+
+    return client;
+}
+
+export async function runEval(
+    client: HumanloopClient,
+    file: File,
+    dataset: Dataset,
+    name?: string,
+    evaluators: Evaluator[] = [],
+    workers: number = 8,
+): Promise<EvaluatorCheck[]> {
+    // Get or create the file on Humanloop
+    if (!file.path && !file.id) {
+        throw new Error("You must provide a path or id in your `file`.");
+    }
+
+    if (file.callable && "path" in file.callable) {
+        if (file.path !== file.callable.path) {
+            throw new Error(
+                "The path of the evaluated `file` must match the path of your decorated `callable`. Expected path: " +
+                    file.path +
+                    ", got: " +
+                    file.callable.path,
+            );
+        }
+    }
+
+    if (file.callable && "version" in file.callable) {
+        if (!_.isEqual(file.version, file.callable.version)) {
```

Comment:
do we want to just check version_id? is a deep equal necessary?

Reply:
version can be a deep object, e.g. `{flow: attributes: {...}}`

Reply:
if both have `version_id`, i think we should use `version_id`.

as noted in other comment:
the deep equal here is slightly concerning - very often the response a user would get from the API won't be exactly equal to the request made. (e.g. the response has default values injected, along with some null attributes)

Reply:
I've been considering it, and I think we should prevent the user from specifying a value for `file.version` when the callable is wrapped in an SDK utility. There's a mini design-doc on this issue here:

https://linear.app/humanloop/issue/ENG-1456/run-utility-callable-version-issue

I'd rather leave this as it is and revisit when we tackle the ticket above.

Reply:
Discussed in a sync with Harry. Was scoped out for later